pandas/_libs/tslibs/src/datetime/np_datetime.c

```diff
@@ -378,7 +378,7 @@ int convert_pydatetime_to_datetimestruct(PyObject *dtobj,
             if (tmp == NULL) {
                 return -1;
             }
-            seconds_offset = PyInt_AsLong(tmp);
+            seconds_offset = PyInt_AsLong(PyNumber_Long(tmp));
```

Comment:
PyNumber_Long returns a new reference, so the way this is currently written it will leak memory. You ideally should break this into a few lines and be sure to call Py_DECREF on the object returned from PyNumber_Long

While touching this might as well also switch from PyInt_AsLong to PyLong_AsLong

Reply:
> While touching this might as well also switch from PyInt_AsLong to PyLong_AsLong
Will try to do this as a follow up as I'm planning to have this PR backported.